feat: Oracle connector enhancements

Dilli-Babu-Godari opened the pull request:
## Description
 Oracle connector enhancements.

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan


```
presto> CREATE TABLE oracle.tm_lakehouse_engine.orcaemp2 (
     ->     empid DOUBLE,
     ->     empname VARCHAR(255),
     ->     empbandcode DOUBLE,
     ->     location DOUBLE,
     ->     empdepcode DOUBLE,
     ->     address VARCHAR(255),
     ->     email VARCHAR(255),
     ->     ssn VARCHAR(255),
     ->     phone VARCHAR(255)
     -> );
CREATE TABLE

Query 20260326_071832_00019_85sxf, FINISHED, 0 nodes
http://localhost:8080/ui/query.html?20260326_071832_00019_85sxf
Splits: 0 total, 0 done (0.00%)
[Latency: client-side: 0:12, server-side: 0:12] [0 rows, 0B] [0 rows/s, 0B/s]


presto> INSERT INTO oracle.tm_lakehouse_engine.orcaemp2
     ->     (empid, empname, empbandcode, location, empdepcode, address, email, ssn, phone)
     -> VALUES
     ->     (1.0, 'John Doe', 3.0, 101.0, 20.0, '123 Main St, New York, NY', 'john.doe@example.com', '123-45-6789', '555-123-4567');

Query 20260326_071911_00020_85sxf, RUNNING, 1 node, 35 splits
http://localhost:8080/ui/query.html?20260326_071911_00020_85sxf

INSERT: 1 row

Query 20260326_071911_00020_85sxf, FINISHED, 1 node
http://localhost:8080/ui/query.html?20260326_071911_00020_85sxf
Splits: 35 total, 35 done (100.00%)
[Latency: client-side: 0:44, server-side: 0:44] [0 rows, 0B] [0 rows/s, 0B/s]


presto> SELECT * FROM oracle.tm_lakehouse_engine.orcaemp2;

Query 20260326_072023_00021_85sxf, RUNNING, 1 node, 17 splits

empid | empname  | empbandcode | location | empdepcode |          address          |        email         |     ssn     |    phone     
-------+----------+-------------+----------+------------+---------------------------+----------------------+-------------+--------------
   1.0 | John Doe |         3.0 |    101.0 |       20.0 | 123 Main St, New York, NY | john.doe@example.com | 123-45-6789 | 555-123-4567 
(1 row)

Query 20260326_072023_00021_85sxf, FINISHED, 1 node
http://localhost:8080/ui/query.html?20260326_072023_00021_85sxf
Splits: 17 total, 17 done (100.00%)
[Latency: client-side: 0:21, server-side: 0:21] [1 rows, 0B] [0 rows/s, 0B/s]
```

**Testing for fetch_size:**

**Presto-CLI:**

_jdbc-fetch-size = 10_

```
presto> select * from oracle_small_fetch.TM_LAKEHOUSE_ENGINE.TEST_FETCH;
    ID    |  DATA   
------------------+------------
 1211.0000000000 | data_1211  
 1212.0000000000 | data_1212  
 1213.0000000000 | data_1213  
 1214.0000000000 | data_1214  
 1215.0000000000 | data_1215  
 1216.0000000000 | data_1216  
 1217.0000000000 | data_1217  
 1218.0000000000 | data_1218  
 1219.0000000000 | data_1219  
 1220.0000000000 | data_1220
(query aborted by user)

Query 20260407_144910_00006_m2vrb, RUNNING, 1 node
Splits: 17 total, 0 done (0.00%)
[Latency: client-side: 1:15, server-side: 1:11] [0 rows, 0B] [0 rows/s, 0B/s]

```
<img width="1144" height="580" alt="Screenshot 2026-04-07 at 9 33 16 PM" src="https://github.com/user-attachments/assets/8a95574a-e17b-43de-a29f-7a9e5ff94161" />


[20260407_144910_00006_m2vrb_small_fetch.json](https://github.com/user-attachments/files/26543492/20260407_144910_00006_m2vrb_small_fetch.json)


_jdbc-fetch-size = 1000_

```
presto> select * from oracle_large_fetch.TM_LAKEHOUSE_ENGINE.TEST_FETCH;
    ID    |  DATA   
------------------+------------
 1211.0000000000 | data_1211  
 1212.0000000000 | data_1212  
 1213.0000000000 | data_1213  
 1214.0000000000 | data_1214  
 1215.0000000000 | data_1215  
 1216.0000000000 | data_1216  
 1217.0000000000 | data_1217  
 1218.0000000000 | data_1218  
 1219.0000000000 | data_1219  
 1220.0000000000 | data_1220  
 1221.0000000000 | data_1221  
 1222.0000000000 | data_1222
(query aborted by user)

Query 20260407_144746_00005_m2vrb, RUNNING, 1 node
Splits: 17 total, 0 done (0.00%)
[Latency: client-side: 0:11, server-side: 0:06] [0 rows, 0B] [0 rows/s, 0B/s]
```
[20260407_144746_00005_m2vrb_large_fetch.json](https://github.com/user-attachments/files/26543501/20260407_144746_00005_m2vrb_large_fetch.json)


<img width="1186" height="469" alt="Screenshot 2026-04-07 at 9 31 40 PM" src="https://github.com/user-attachments/assets/5defb615-f3bb-4627-940d-fc222c13bdc7" />

## Contributor checklist

- [ ] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

Oracle Connector Changes
*  Add oraorai18n.jar dependency, jdbc fetchsize and SSL config for Oracle connector.

agrawalreetika commented on presto-oracle/src/main/java/com/facebook/presto/plugin/oracle/OracleConfig.java:
```diff
@@ -95,4 +100,64 @@ public OracleConfig setTimestampDefaultPrecision(int timestampDefaultPrecision)
+        return this;
+    }
+
+    public String getKeystorePath()
```
Where are we using/setting this?

Dilli-Babu-Godari replied:
we are not using it anywhere as of now.

agrawalreetika commented on presto-oracle/src/main/java/com/facebook/presto/plugin/oracle/OracleConfig.java:
```diff
@@ -95,4 +100,64 @@ public OracleConfig setTimestampDefaultPrecision(int timestampDefaultPrecision)
+        return this;
+    }
+
+    public String getKeystorePassword()
```
Where are we using/setting this?

agrawalreetika reviewed (commented):
I have a few comments - 
1. Let's split PR in 2 commits - TLS & fetchsize changes
2. I think TLS is going to be a generic feature acorss different JDBC connectors, maybe we can create a separate TLS config or have it in BaseJDBC config and reuse it in the required JDBC connector?

Dilli-Babu-Godari committed:
Add fetch_size for oracle getTables to improve the performance

Oracle connector fails with a connection timeout, while show tables, when the schema has a 1 lakh tables. This happens because we are not specifying the fetch_size with the getTables method in Oracle. As a result, the connector uses the default fetch_size, which is a smaller value (10), leading to multiple network calls to the datasource and ultimately causing a connection timeout.,

Co-authored-by: lukmanulhakkeem <lukmanul.hakkeem.a@ibm.com>

Dilli-Babu-Godari committed:
Adds orai18n.jar to ensure Oracle works with charsets like WE8ISO8859P9.

added the orai18n.jar dependency, ensuring compatibility with non-UTF character sets like WE8ISO8859P9.

Dilli-Babu-Godari committed:
Added ssl configs for oracle

SSL enablement of Oracle Connector with Oracle THIN JDBC driver

Co-authored-by: lukmanulhakkeem <lukmanul.hakkeem.a@ibm.com>

agrawalreetika commented on presto-oracle/pom.xml:
```diff
@@ -195,6 +195,13 @@
+            <artifactId>orai18n</artifactId>
+            <version>19.3.0.0</version>
+            <scope>runtime</scope>
+        </dependency>
```
What is this for?

Dilli-Babu-Godari commented:
> > Updated the changes
> 
> Thanks for addressing the last point! Please address these:
> 
> * The PR is now automatically included, you don't have to add it in yourself anymore.
> * Revise the entry to begin with a word from the [Order of changes](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#order-of-changes) in the Release Notes Guidelines: is the work in this PR a Fix, for example?
> * Expand the abbreviation WxD.

Updated the changes.

steveburnett commented:
Formatting nits and suggestions for the release note: 

```
== RELEASE NOTES ==

Oracle Connector Changes
*  Add oraorai18n.jar dependency and SSL config for Oracle connector.
```

agrawalreetika reviewed (changes requested):
Currently - `Added ssl configs for oracle` commit needs refactoring there are multiple unused configurations there. Please check

Dilli-Babu-Godari commented on presto-oracle/src/main/java/com/facebook/presto/plugin/oracle/OracleConfig.java:
```diff
@@ -95,4 +100,64 @@ public OracleConfig setTimestampDefaultPrecision(int timestampDefaultPrecision)
         return this;
     }
+
+    public boolean isTlsEnabled()
```
I plan to use these methods in a follow-up PR but to keep this one streamlined, I'll remove them here.

Updated commit message as well.

Dilli-Babu-Godari commented:
> Formatting nits and suggestions for the release note:
> 
> ```
> == RELEASE NOTES ==
> 
> Oracle Connector Changes
> *  Add oraorai18n.jar dependency and SSL config for Oracle connector.
> ```

Updated changes.

agrawalreetika reviewed (changes requested):
I think we should have one more flag -`oracle.tls.enabled` and then based on that add it to config.
Also what testing is done for this? I don't see any tests cases as well?

agrawalreetika reviewed (changes requested):
I don't see any usage of `isTlsEnabled`? Also commit config tests in existing resptective commits, it doesn;t require a separate commit

Dilli-Babu-Godari commented:
> I don't see any usage of `isTlsEnabled`? Also commit config tests in existing resptective commits, it doesn;t require a separate commit

Changes done, let me know if anything else is needed.

Dilli-Babu-Godari commented on presto-base-jdbc/src/test/java/com/facebook/presto/plugin/jdbc/TestBaseJdbcConfig.java:
```diff
@@ -67,7 +70,8 @@ public void testExplicitPropertyMappings()
                 .setCaseInsensitiveNameMatchingCacheTtl(new Duration(1, SECONDS))
-                .setCaseSensitiveNameMatching(true);
+                .setCaseSensitiveNameMatching(true)
+                .setFetchSize(5000);
```
Included changes in the respective commits.

agrawalreetika commented:
Please add test cases for your PR as well, once Oracle tests are enabled.

agrawalreetika reviewed (changes requested):
Thanks for the PR. 
Please add documentation for all the newly added properties.

Dilli-Babu-Godari committed:
Add TLS documentation for Oracle connector

steveburnett reviewed (changes requested):
A local doc build returns the following error:

`/Users/steveburnett/Documents/GitHub/presto/presto-docs/src/main/sphinx/connector/oracle.rst:56: ERROR: Malformed table.
Bottom/header table border does not match top border.`

The entire table in General Configuration Properties is not displayed as a result. See screenshot of my local doc build. 

<img width="625" height="340" alt="Screenshot 2026-04-01 at 9 42 55 AM" src="https://github.com/user-attachments/assets/b9beab91-309a-4fb2-839c-ab787910b529" />



Please fix the formatting of the table and re-request my review, and I'll be happy to review the content.

steveburnett reviewed (dismissed):
LGTM! (docs) 

Pull updated branch, new local doc build, looks good. Thanks!

Dilli-Babu-Godari committed:
Revert custom ConnectionFactory to set Oracle's default row prefetch

agrawalreetika commented on presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/BaseJdbcConfig.java:
```diff
@@ -191,4 +192,16 @@ public void validateConfig()
+    {
+        this.fetchSize = fetchSize;
+        return this;
+    }
```
Can you link the issue to this, if you have it?

Dilli-Babu-Godari replied:
This is the Issue created to track this - https://github.com/prestodb/presto/issues/27536

agrawalreetika reviewed (changes requested):
Mostly lgtm apart from few comments.

agrawalreetika commented on presto-oracle/src/main/java/com/facebook/presto/plugin/oracle/OracleClient.java:
```diff
@@ -116,19 +117,21 @@ protected ResultSet getTables(Connection connection, Optional<String> schemaName
                 escapeNamePattern(schemaName, Optional.of(escape)).orElse(null),
                 escapeNamePattern(tableName, Optional.of(escape)).orElse(null),
                 getTableTypes());
+        resultSet.setFetchSize(fetchSize);
```
So I looked into Oracle's default for `defaultRowPrefetch`. In your current implementation, setting `fetchSize` after `metadata.getTables()` executes means the first fetch still uses Oracle's default of 10 rows. So setting fetchSize afterwards only affects subsequent fetches, not the first one. 

Despite the suboptimal first fetch, this implementation solves the main problem by reducing network calls based on the configured `jdbc-fetch-size` and keeping the first fetch size as 10.

This is Ok and solves the problem, but not very clean, I would say.
Other ways to optimize fetchsize across all the operations could be by setting setting `defaultRowPrefetch` at the connection level in `OracleClientModule`: -
```
connectionProperties.setProperty("defaultRowPrefetch", fetchSize);
```

steveburnett reviewed (approved):
LGTM! (docs) 

Pull branch, local doc build, looks good. Thanks!

tdcmeehan commented:
Can this be split into 3 different PRs, one for each of the Oracle enhancements?

Dilli-Babu-Godari commented:
> Can this be split into 3 different PRs, one for each of the Oracle enhancements?

sure @tdcmeehan will do that and update you!

Dilli-Babu-Godari commented:
Closing this PR since we broke down this into multiple sub PRs

https://github.com/prestodb/presto/pull/27671
https://github.com/prestodb/presto/pull/27670
https://github.com/prestodb/presto/pull/27669
